### ISSUE TITLE
default テンプレートをレスポンシブWebデザイン対応する

### ISSUE THREAD

**Issue by seasoftjapan** [EC-CUBE/ec-cube2#1065]
スマートフォン端末のテンプレートを default テンプレートに設定した際に、それなりに使える状態を目指す。

- 既存のテンプレートをPC・スマートフォンで別々に設定できる仕組みは残す。

#1057 から派生。

**Comment by seasoftjapan** [comment]
奇跡的に、2015 年以降、#804 以外は、一切手が加わっていないようです。
```
$ git log -3 -- html/user_data/packages/default/style.css html/user_data/packages/default/css/
commit 84e65a713289ced15b856ec6c10168bfd7636284
Author: Seasoft <info@seasoft.jp.net>
Date:   Wed Dec 20 19:50:58 2023 +0900

    CSS Nesting Module を使った実装 #783 (PC)

    目視で気になった点を調整した。

commit e332637fa75d4552d0ab589195bf0cced68ab3e9
Author: Seasoft <info@seasoft.jp.net>
Date:   Wed Dec 20 18:56:59 2023 +0900

    CSS #783 #784 (PC)

    - CSS Nesting Module を使った実装 #783
    - CSS ファイルをまとめる #784

    ```
    cat \
    html/user_data/packages/default/css/reset.css \
    html/user_data/packages/default/css/common.css \
    html/user_data/packages/default/css/contents.css \
    html/user_data/packages/default/css/table.css \
    html/user_data/packages/default/css/bloc.css \
    html/user_data/packages/default/css/bloc_alpha.css \
    html/user_data/packages/default/css/popup.css \
    html/user_data/packages/default/css/print.css \
    | sass-convert --indent=4 --from=css --to=scss \
    > html/user_data/packages/default/style.css
    ```

commit 949b4ea1c5ed41e9d62cd04b268c3d222084d5d6
Author: miyamura <miyamura@42d9b83e-2207-4a45-8b47-68c1da84f352>
Date:   Sun Sep 14 07:02:29 2014 +0000

    #2617 デフォルトデザインテンプレートのデザイン崩れ修正対応
```

**Comment by nanasess** [comment]
@seasoftjapan そうですね。フロント側のテンプレートは、不具合修正以外はメジャーバージョンアップでしか変更しないポリシーだったので、ほとんど変更無いと思います

**Comment by seasoftjapan** [comment]
ライセンス周りの確認です。 > 分かる方

https://github.com/EC-CUBE/ec-cube2/issues/1057#issuecomment-2477733290
> @seasoftjapan 以前（かなり前に）作ったテンプレート使えたらいいんですけど・・笑 https://www.dropbox.com/scl/fo/pwlzyndqennu56i405hgx/AALFIXH520gvqbkb9MEN5uU?rlkey=tkunz66hjuvtkb9mh003yp29g&dl=0
> 
> かなり修正は要ると思いますが、これでよければ好きに使っていただいて大丈夫です 使えればですが笑

これを取り込む流れですが、私がソース取得してコミットして大丈夫ですか?
外谷さんから PR いただく流れが必要ですか?

**Comment by bluestylejp-toya** [comment]
私は問題ありません！

**Comment by bluestylejp-toya** [comment]
@seasoftjapan こちらもご利用ください！
https://www.dropbox.com/home/bluestylejp/04_EC-CUBE/2.17%E3%83%AC%E3%82%B9%E3%83%9D%E3%83%B3%E3%82%B7%E3%83%96%E5%8C%96

**Comment by seasoftjapan** [comment]
> @seasoftjapan こちらもご利用ください！ https://www.dropbox.com/home/bluestylejp/04_EC-CUBE/2.17%E3%83%AC%E3%82%B9%E3%83%9D%E3%83%B3%E3%82%B7%E3%83%96%E5%8C%96

ありがとうございます。
後ほど、取り込んだものを Push させていただきます。

**Comment by seasoftjapan** [comment]
レジュー用に差分をわかりやすくするため、#804 適用前でブランチを作成しました。
https://github.com/EC-CUBE/ec-cube2/compare/seasoft-1065

data/Smarty/templates/admin/design/subnavi.tpl
data/class/SC_Display.php
の変更は、冒頭で触れた「既存のテンプレートをPC・スマートフォンで別々に設定できる仕組み」を無効化する内容で、要検討と認識しています。

もしも、無効化すべきなどの意見や、両立の良いアイディアがありましたらコメントください。

個人的には、スマートフォンでも default テンプレートを選択可能として、選択した場合の動作を調整すれば両立できそうと思っているのですが、実際に試すと課題もありそうでした。

**Comment by seasoftjapan** [comment]
ちなみに、スマホも default テンプレート (レスポンシブWebデザイン対応) を使うけど、ブロックの割当は PC と別にしたい (「デザイン管理＞スマートフォン＞レイアウト設定」画面を使いたい) 需要ってありますか?

**Comment by seasoftjapan** [comment]
> 個人的には、スマートフォンでも default テンプレートを選択可能として、選択した場合の動作を調整すれば両立できそうと思っているのですが、実際に試すと課題もありそうでした。

スマートフォンでも default テンプレートを選択可能として、選択した場合に PC のレイアウト情報を読み込むように調整しました。[8f643f0](https://github.com/EC-CUBE/ec-cube2/commit/8f643f06219a0c1bd97472a13713393755420fe9)

> ちなみに、スマホも default テンプレート (レスポンシブWebデザイン対応) を使うけど、ブロックの割当は PC と別にしたい (「デザイン管理＞スマートフォン＞レイアウト設定」画面を使いたい) 需要ってありますか?

需要が未知な割に面倒も多いため、対応していません。

**Comment by seasoftjapan** [comment]
備忘録: MySQL テスト落ちているっぽい。そういえば、MySQL のテーブル定義変更してない。後でやる。

**Comment by nanasess** [comment]
> ちなみに、スマホも default テンプレート (レスポンシブWebデザイン対応) を使うけど、ブロックの割当は PC と別にしたい (「デザイン管理＞スマートフォン＞レイアウト設定」画面を使いたい) 需要ってありますか?

ブロックの割り当てを別にしたい場合は、レスポンシブテンプレートを諦めてもらって、従来の pc(default)/sphone を利用していただければよいと思われます

**Comment by seasoftjapan** [comment]
> ブロックの割り当てを別にしたい場合は、レスポンシブテンプレートを諦めてもらって、従来の pc(default)/sphone を利用していただければよいと思われます

はい。sphone は従来通り使えます。(そもそも標準は sphone のまま。)

sphone との互換性を維持しつつ対応しようと思うと、dtb_pagelayout, dtb_bloc, dtb_blocposition 辺りは template_code 列を追加(テーブルによっては変更?)して、主キーに組み入れないと厳しいかなという印象でした。ただ、それ決め打ちだと、ブロックの割当は PC と共通にしたいサイトにとって面倒になるので、そこのカバーも必要そうですね (パラメーターとかで実現できそうではありますが)。

個別サイトのカスタマイズとしては、sphone との互換性を破棄すれば、それほど難なく対応できると思います。

あとは、ブロック割当共通をゴリ押しできそうな要素として、テンプレート内で `<!--{if SC_Display_Ex::detectDevice() === DEVICE_TYPE_SMARTPHONE}-->` とかできます。テンプレートに関わらず、端末種別を返します。

実際、CSS (メディアクエリ width) 判定の画面表示と detectDevice() の結果が矛盾して、意図しない組み合わせになる懸念もあるので、PC/SP の判別をメディアクエリに頼らず detectDevice() の結果に従うよう調整する必要もあるかもしれません。(CSS nesting との相性も良くレスポンシブの書き分けも柔軟になります。それを標準にしても良いとも思いますが、色々な端末が存在するので、どっちが優位とも言い難いところです。)

**Comment by nanasess** [comment]
@seasoftjapan 
既存の2.13や2.17からのバージョンアップを考慮して、 default テンプレートを非レスポンシブのまま使用することはできそうですかね？

**Comment by seasoftjapan** [comment]
> @seasoftjapan 既存の2.13や2.17からのバージョンアップを考慮して、 default テンプレートを非レスポンシブのまま使用することはできそうですかね？

具体的なシナリオにより回答は変わります。

例えば、PCブラウザーでウィンドウ幅を狭めた場合には、レスポンシブ動作となる認識です。標準実装として回避するには、上で触れた「PC/SP の判別をメディアクエリに頼らず detectDevice() の結果に従うよう調整する」という手法を適用すると解決できると思います。

そうではなく、新版に対してレスポンシブ対応前のテンプレートを適用した場合、従来通り非レスポンシブの動作となる認識です。

**Comment by nanasess** [comment]
@seasoftjapan 
ありがとうございます！
`TEMPLATE_NAME` のデフォルト値は `default` で、 `sphone` に変更するとレスポンシブ動作になるという認識であってますかね？

**Comment by seasoftjapan** [comment]
> @seasoftjapan ありがとうございます！ `TEMPLATE_NAME` のデフォルト値は `default` で、 `sphone` に変更するとレスポンシブ動作になるという認識であってますかね？

`TEMPLATE_NAME` のデフォルト値は `default` です。これはそのままの想定です。

`SMARTPHONE_TEMPLATE_NAME` が `default` になると、スマートフォン端末にも PC テンプレートが反映され、結果的に同一テンプレートでレスポンシブ動作となる認識です。
